### PR TITLE
feat(transfer)!: migrate transfer parameters to be self managed

### PR DESCRIPTION
**Pull request by srdtrk**
## Description

With this PR:

- [x] The `transfer` app now self-manages its parameters.
- [x] Unit and integration tests have been added for the `transfer` app's parameters.
- [x] Migration code has been implemented for moving parameters from `x/params`' subspace to the `transfer` app's keeper.
- [x] Migration unit and integration tests have been added.
- [x] End to end tests have been updated.

closes: #3502

addresses: #2010 

### Changes:

- A new message type, `MsgUpdateParams`, has been introduced, executed with an authority that is expected to be the `x/gov` module account.

### Commit Message / Changelog Entry

```
feat(transfer)!: migrate transfer parameters to be self managed
```

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) for commit messages. (view raw markdown for examples)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [x] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [x] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [x] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [x] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [x] Re-reviewed `Files changed` in the Github PR explorer.
- [x] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: e2e/tests/transfer/base_test.go [hunk 1]
```diff
@@ -274,12 +283,17 @@ func (s *TransferTestSuite) TestSendEnabledParam() {
 	})
 
 	t.Run("change send enabled parameter to disabled", func(t *testing.T) {
-		changes := []paramsproposaltypes.ParamChange{
-			paramsproposaltypes.NewParamChange(transfertypes.StoreKey, string(transfertypes.KeySendEnabled), "false"),
-		}
+		if isSelfManagingParams {
```
Comment (damiannolan) [inline e2e/tests/transfer/base_test.go:286]: You should be able to use a [`FeatureReleaases`](https://github.com/cosmos/ibc-go/blob/main/e2e/semverutil/semver.go#L10) matrix to do custom versioning here. 

Here is an example of how its [defined](https://github.com/cosmos/ibc-go/blob/main/e2e/tests/transfer/base_test.go#L400-L409) and [used in a test](https://github.com/cosmos/ibc-go/blob/main/e2e/tests/transfer/base_test.go#L442-L447)

**Review by damiannolan — COMMENTED** [review]
Great work so far! I left various suggestions, I will continue to review tomorrow.

Its interesting to note that `x/params` used to maintain its key tables in memory, and the keepers had reference to a `Subspace` which held those key tables, right? 
So now we are reading from the store each time there is a param lookup. It looks like the sdk may be handling this handling this here - https://github.com/cosmos/cosmos-sdk/blob/main/x/auth/keeper/keeper.go#L79

File: modules/apps/transfer/keeper/msg_server.go [hunk 2]
```diff
@@ -5,18 +5,32 @@ import (
...
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	govtypes "github.com/cosmos/cosmos-sdk/x/gov/types"
 
 	ibcerrors "github.com/cosmos/ibc-go/v7/internal/errors"
 	"github.com/cosmos/ibc-go/v7/modules/apps/transfer/types"
 )
 
 var _ types.MsgServer = (*Keeper)(nil)
 
+func (k Keeper) UpdateParams(goCtx context.Context, msg *types.MsgUpdateParams) (*types.MsgUpdateParamsResponse, error) {
```
Comment (damiannolan) [inline modules/apps/transfer/keeper/msg_server.go:16]: could we move `UpdateParams` to below `Transfer` at the bottom of the file

File: modules/apps/transfer/types/codec.go [hunk 3]
```diff
@@ -29,6 +29,11 @@ func RegisterInterfaces(registry codectypes.InterfaceRegistry) {
 		&TransferAuthorization{},
 	)
 
+	registry.RegisterImplementations(
+		(*sdk.Msg)(nil),
+		&MsgUpdateParams{},
+	)
+
```
[the remark this answers is not in the document]
Comment (srdtrk) [reply modules/apps/transfer/types/codec.go]: yes, implemented

File: modules/apps/transfer/keeper/msg_server_test.go [hunk 4]
```diff
@@ -125,3 +139,47 @@ func (suite *KeeperTestSuite) TestMsgTransfer() {
 		})
 	}
 }
+
+// TestUpdateParams tests UpdateParams rpc handler
+func (suite *KeeperTestSuite) TestUpdateParams() {
+	validAuthority := suite.chainA.GetSimApp().TransferKeeper.GetAuthority()
+	testCases := []struct {
+		name    string
+		request *types.MsgUpdateParams
```
[the remark this answers is not in the document]
Comment (srdtrk) [reply modules/apps/transfer/keeper/msg_server_test.go]: implemented

File: modules/apps/transfer/keeper/msg_server.go (hunk 2 above)
Comment (srdtrk) [reply modules/apps/transfer/keeper/msg_server.go]: implemented

File: modules/apps/transfer/keeper/msg_server_test.go [hunk 5]
```diff
@@ -1,13 +1,26 @@
...
+// define constants used for testing
+const (
+	invalidAddress = "invalid"
+)
+
+var (
+	emptyAddr string
+
+	validAddress = sdk.AccAddress(secp256k1.GenPrivKey().PubKey().Address()).String()
+)
```
[the remark this answers is not in the document]
Comment (srdtrk) [reply modules/apps/transfer/keeper/msg_server_test.go]: implemented

File: e2e/tests/transfer/base_test.go (hunk 1 above)
Comment (srdtrk) [reply e2e/tests/transfer/base_test.go:275]: implemented

File: modules/apps/transfer/keeper/msg_server.go [hunk 6]
```diff
@@ -5,18 +5,32 @@ import (
...
+
+	return &types.MsgUpdateParamsResponse{}, nil
+}
+
 // Transfer defines an rpc handler method for MsgTransfer.
 func (k Keeper) Transfer(goCtx context.Context, msg *types.MsgTransfer) (*types.MsgTransferResponse, error) {
 	ctx := sdk.UnwrapSDKContext(goCtx)
 
-	if !k.GetSendEnabled(ctx) {
+	if !k.GetParams(ctx).SendEnabled {
```
[the remark this answers is not in the document]
Comment (srdtrk) [reply modules/apps/transfer/keeper/msg_server.go:19]: I slightly don't but I'd also like to hear what others thing. I like it when every get also has a set. And params should be set together

File: modules/apps/transfer/types/msgs_test.go [hunk 7]
```diff
@@ -38,6 +40,9 @@ var (
 	zeroCoin         = sdk.Coin{Denom: "atoms", Amount: sdk.NewInt(0)}
 
 	timeoutHeight = clienttypes.NewHeight(0, 10)
+
+	validAuthority = sdk.AccAddress("authority").String()
```
Comment (crodriguezvega) [inline modules/apps/transfer/types/msgs_test.go:44]: Sorry, maybe I'm missing something but how can this be a valid authority address?

**Review by crodriguezvega — COMMENTED** [review]
Thanks for working on this, @srdtrk!

File: modules/apps/transfer/keeper/keeper.go [hunk 8]
```diff
@@ -22,47 +21,64 @@ import (
 
 // Keeper defines the IBC fungible transfer keeper
 type Keeper struct {
-	storeKey   storetypes.StoreKey
-	cdc        codec.BinaryCodec
-	paramSpace paramtypes.Subspace
+	storeKey       storetypes.StoreKey
+	cdc            codec.BinaryCodec
+	legacySubspace paramtypes.Subspace
```
[the remark this answers is not in the document]
Comment (srdtrk) [reply modules/apps/transfer/keeper/keeper.go:26]: Originally, it was like the SDK. We had a large discussion with damian and colin over on slack, and chose to take this route

[the remark this answers is not in the document]
Comment (crodriguezvega) [reply modules/apps/transfer/keeper/keeper.go:26]: Cool, I only skimmed thought the discussion so missed the details :) 👍

File: docs/apps/transfer/params.md [hunk 9]
```diff
@@ -28,3 +34,57 @@ To prevent a single token from being transferred to the chain, set the `ReceiveE
 
 - For Cosmos SDK v0.46.x or earlier, set the bank module's [`SendEnabled` parameter](https://github.com/cosmos/cosmos-sdk/blob/release/v0.46.x/x/bank/spec/05_params.md#sendenabled) for the denomination to `false`.
 - For Cosmos SDK versions above v0.46.x, set the bank module's `SendEnabled` entry for the denomination to `false` using `MsgSetSendEnabled` as a governance proposal.
+
+::: warning
+Doing so will prevent the token from being transferred between any accounts in the blockchain.
+:::
+
+## Queries
```
[the remark this answers is not in the document]
Comment (srdtrk) [reply docs/apps/transfer/params.md:42]: In another docs related PR?

File: modules/apps/transfer/types/msgs_test.go (hunk 7 above)
Comment (srdtrk) [reply modules/apps/transfer/types/msgs_test.go:44]: It's not valid, that variable is unused and a mistake, thanks for catching, instead, I redefined it down the line

Comment (srdtrk) [reply modules/apps/transfer/types/msgs_test.go:44]: Yeah, doesn't make much sense. Test passes only because it is wrapped by `AccAddress`. I took this test case from SDK I believe, I can remove it if you'd like.

File: modules/apps/transfer/keeper/msg_server.go [hunk 10]
```diff
@@ -59,3 +59,14 @@ func (k Keeper) Transfer(goCtx context.Context, msg *types.MsgTransfer) (*types.
 
 	return &types.MsgTransferResponse{Sequence: sequence}, nil
 }
+
+func (k Keeper) UpdateParams(goCtx context.Context, msg *types.MsgUpdateParams) (*types.MsgUpdateParamsResponse, error) {
+	if k.authority != msg.Authority {
+		return nil, errorsmod.Wrapf(ibcerrors.ErrInvalidAddress, "invalid authority; expected %s, got %s", k.authority, msg.Authority)
```
[the remark this answers is not in the document]
Comment (srdtrk) [reply modules/apps/transfer/keeper/msg_server.go:65]: I like `;`, it is taken from the SDK. Both should be acceptable?

File: modules/apps/transfer/types/msgs.go [hunk 11]
```diff
@@ -11,6 +11,38 @@ import (
...
+		Authority: authority,
+		Params:    params,
+	}
+}
+
+// ValidateBasic implements sdk.Msg
+func (msg MsgUpdateParams) ValidateBasic() error {
+	_, err := sdk.AccAddressFromBech32(msg.Authority)
+	if err != nil {
+		return err
```
[the remark this answers is not in the document]
Comment (srdtrk) [reply modules/apps/transfer/types/msgs.go]: done

File: modules/apps/transfer/keeper/keeper_test.go [hunk 12]
```diff
@@ -210,3 +215,53 @@ func (suite *KeeperTestSuite) TestGetAllDenomEscrows() {
...
 }
+
+func (suite *KeeperTestSuite) TestParams() {
+	testCases := []struct {
+		name    string
+		input   types.Params
+		expPass bool
+	}{
+		// it is not possible to set invalid booleans
+		{name: "success: set params false-false", input: types.NewParams(false, false), expPass: true},
```
[the remark this answers is not in the document]
Comment (srdtrk) [reply modules/apps/transfer/keeper/keeper_test.go]: done

File: modules/apps/transfer/keeper/msg_server_test.go [hunk 13]
```diff
@@ -125,3 +126,48 @@ func (suite *KeeperTestSuite) TestMsgTransfer() {
...
+func (suite *KeeperTestSuite) TestUpdateParams() {
+	var emptyAddr string
+	validAuthority := suite.chainA.GetSimApp().TransferKeeper.GetAuthority()
+	testCases := []struct {
+		name    string
+		msg     *types.MsgUpdateParams
+		expPass bool
+	}{
+		{
+			name:    "success",
```
[the remark this answers is not in the document]
Comment (srdtrk) [reply modules/apps/transfer/keeper/msg_server_test.go]: done

File: modules/apps/transfer/keeper/msg_server_test.go [hunk 14]
```diff
@@ -125,3 +126,48 @@ func (suite *KeeperTestSuite) TestMsgTransfer() {
...
+			msg:     types.NewMsgUpdateParams(validAuthority, types.DefaultParams()),
+			expPass: true,
+		},
+		{
+			name:    "invalid authority address",
+			msg:     types.NewMsgUpdateParams(ibctesting.InvalidID, types.DefaultParams()),
```
[the remark this answers is not in the document]
Comment (srdtrk) [reply modules/apps/transfer/keeper/msg_server_test.go]: done

File: e2e/tests/transfer/base_test.go [hunk 15]
```diff
@@ -29,34 +28,25 @@ type TransferTestSuite struct {
 	testsuite.E2ETestSuite
 }
 
+// transferSelfParamFeatureReleases represents the releases the transfer module started managing its own params.
+var transferSelfParamFeatureReleases = semverutil.FeatureReleases{
```
[the remark this answers is not in the document]
Comment (srdtrk) [reply e2e/tests/transfer/base_test.go]: done

File: modules/apps/transfer/keeper/keeper.go [hunk 16]
```diff
@@ -93,6 +109,25 @@ func (k Keeper) SetPort(ctx sdk.Context, portID string) {
 	store.Set(types.PortKey, []byte(portID))
 }
 
+// GetParams returns the current transfer module parameters.
+func (k Keeper) GetParams(ctx sdk.Context) (p types.Params) {
```
Comment (damiannolan) [inline modules/apps/transfer/keeper/keeper.go:113]: nit: remove named return arg `p`

File: modules/apps/transfer/keeper/msg_server.go [hunk 17]
```diff
@@ -59,3 +59,14 @@ func (k Keeper) Transfer(goCtx context.Context, msg *types.MsgTransfer) (*types.
 
 	return &types.MsgTransferResponse{Sequence: sequence}, nil
 }
+
+func (k Keeper) UpdateParams(goCtx context.Context, msg *types.MsgUpdateParams) (*types.MsgUpdateParamsResponse, error) {
```
Comment (damiannolan) [inline modules/apps/transfer/keeper/msg_server.go:63]: needs godoc

File: modules/apps/transfer/keeper/msg_server_test.go [hunk 18]
```diff
@@ -125,3 +126,48 @@ func (suite *KeeperTestSuite) TestMsgTransfer() {
 		})
 	}
 }
+
+// TestUpdateParams tests UpdateParams rpc handler
+func (suite *KeeperTestSuite) TestUpdateParams() {
+	var emptyAddr string
```
Comment (damiannolan) [inline modules/apps/transfer/keeper/msg_server_test.go:132]: Is there any need for this var? Can we just use a string literal below in the testcase?

File: modules/apps/transfer/types/genesis.go [hunk 19]
```diff
@@ -35,8 +35,5 @@ func (gs GenesisState) Validate() error {
 	if err := gs.DenomTraces.Validate(); err != nil {
 		return err
 	}
-	if err := gs.Params.Validate(); err != nil {
-		return err
-	}
```
Comment (damiannolan) [inline modules/apps/transfer/types/genesis.go:40]: how come this is being removed?

File: modules/apps/transfer/types/msgs_test.go [hunk 20]
```diff
@@ -98,3 +103,36 @@ func TestMsgTransferGetSigners(t *testing.T) {
...
+
+// TestMsgUpdateParamsValidation tests ValidateBasic for MsgUpdateParams
+func TestMsgUpdateParamsValidation(t *testing.T) {
+	testCases := []struct {
+		name    string
+		msg     *types.MsgUpdateParams
+		expPass bool
+	}{
+		{"valid authority with valid params", types.NewMsgUpdateParams(validAuthority, types.DefaultParams()), true},
+		{"valid authority address with valid params", types.NewMsgUpdateParams(validAddress, types.DefaultParams()), true},
```
Comment (damiannolan) [inline modules/apps/transfer/types/msgs_test.go:115]: I guess these testcases are more or less the same. We could remove `validAuthority` and `validAddress` and just use `ibctesting.TestAccAddress`

**Review by damiannolan — COMMENTED** [review]
Last few nits from me, looking really good!

File: modules/apps/transfer/types/genesis.go (hunk 19 above)
Comment (srdtrk) [reply modules/apps/transfer/types/genesis.go:40]: Transfer Params no longer have a `Validate` function. Every param in our case is a valid param since they are just booleans. Including a validate function would only reduce codecov since it's function signature would include an error. (The function itself would just say `return nit`)

Comment (damiannolan) [reply modules/apps/transfer/types/genesis.go:40]: makes sense!

Comment (srdtrk) [reply modules/apps/transfer/types/genesis.go:40]: The only reason a validate function existed before was that since the params were managed outside the transfer module, we needed to ensure `x/params` wasn't doing something funny by making sure the returned params were indeed booleans

File: modules/apps/transfer/types/msgs_test.go (hunk 20 above)
Comment (srdtrk) [reply modules/apps/transfer/types/msgs_test.go]: done

File: modules/apps/transfer/keeper/msg_server_test.go (hunk 18 above)
Comment (srdtrk) [reply modules/apps/transfer/keeper/msg_server_test.go]: done

File: modules/apps/transfer/keeper/keeper.go (hunk 16 above)
Comment (srdtrk) [reply modules/apps/transfer/keeper/keeper.go]: done

File: e2e/tests/transfer/base_test.go [hunk 21]
```diff
@@ -29,34 +29,25 @@ type TransferTestSuite struct {
 	testsuite.E2ETestSuite
 }
 
+// transferSelfParamsFeatureReleases represents the releases the transfer module started managing its own params.
+var transferSelfParamsFeatureReleases = semverutil.FeatureReleases{
+	MajorVersion: "v8",
+}
```
Comment (damiannolan) [inline e2e/tests/transfer/base_test.go:35]: feature release matrices are being grouped together in test values in #3558, just noting this for resolving merge conflicts

**Review by damiannolan — APPROVED** [review]
Thanks @srdtrk, nice work! LGTM

File: e2e/tests/transfer/base_test.go [hunk 22]
```diff
@@ -29,34 +29,25 @@ type TransferTestSuite struct {
...
+
 // QueryTransferSendEnabledParam queries the on-chain send enabled param for the transfer module
 func (s *TransferTestSuite) QueryTransferSendEnabledParam(ctx context.Context, chain ibc.Chain) bool {
```
Comment (chatton) [inline e2e/tests/transfer/base_test.go:38]: I wonder should we add a new test function here `QueryTransferParams`

and update the functions to just look something like
```golang
return s.QueryTransferParams(...).SendEnabled
```

what do you think?

File: modules/apps/transfer/keeper/msg_server_test.go [hunk 23]
```diff
@@ -125,3 +126,47 @@ func (suite *KeeperTestSuite) TestMsgTransfer() {
...
+		},
+		{
+			"malformed authority address",
+			types.NewMsgUpdateParams(ibctesting.InvalidID, types.DefaultParams()),
+			false,
+		},
+		{
+			"empty authority address",
+			types.NewMsgUpdateParams("", types.DefaultParams()),
+			false,
```
Comment (chatton) [inline modules/apps/transfer/keeper/msg_server_test.go:151]: maybe we should add a test for a string of whitespace `"           "` also failing here.

**Review by chatton — APPROVED** [review]
Great job with this! Just a few small comments but nothing major LGTM

File: modules/apps/transfer/keeper/msg_server_test.go (hunk 23 above)
Comment (srdtrk) [reply modules/apps/transfer/keeper/msg_server_test.go]: done

File: e2e/tests/transfer/base_test.go (hunk 22 above)
Comment (srdtrk) [reply e2e/tests/transfer/base_test.go]: done
